src/archive/src/main.rs

```diff
@@ -0,0 +1,469 @@
+    static ANCHOR_INDEX: RefCell<AnchorIndex> = with_memory_manager(|memory_manager| {
+        RefCell::new(StableBTreeMap::init(memory_manager.get(ANCHOR_ACCESS_INDEX_MEMORY_ID), std::mem::size_of::<AnchorIndexKey>() as u32, 0))
+    });
+}
+
+/// Reserve the first stable memory page for the configuration stable cell.
+fn config_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 0..1)
+}
+
+/// All the memory after the initial config page is managed by the [MemoryManager].
+fn managed_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 1..MAX_WASM_PAGES)
+}
+
+/// A helper function to access the configuration.
+fn with_config<R>(f: impl FnOnce(&ArchiveConfig) -> R) -> R {
+    CONFIG.with(|cell| f(&cell.borrow().get().get()))
+}
+
+/// A helper function to access the memory manager.
+fn with_memory_manager<R>(f: impl FnOnce(&MemoryManager<Memory>) -> R) -> R {
+    MEMORY_MANAGER.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the log.
+fn with_log<R>(f: impl FnOnce(&StableLog) -> R) -> R {
+    LOG.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the anchor-based index.
+fn with_anchor_index_mut<R>(f: impl FnOnce(&mut AnchorIndex) -> R) -> R {
+    ANCHOR_INDEX.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+/// Configuration state of the archive.
+enum ConfigState {
+    Uninitialized, // This state is only used between wasm module initialization and init().
+    Initialized(ArchiveConfig),
+}
+
+impl ConfigState {
+    fn get(&self) -> &ArchiveConfig {
+        match &self {
+            ConfigState::Uninitialized => trap("archive config not initialized"),
+            ConfigState::Initialized(config) => config,
+        }
+    }
+}
+
+/// Configuration of the archive canister.
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct ArchiveConfig {
+    /// This canister will accept entries only from this principal.
+    ii_canister: Principal,
+    /// The maximum number of entries returned in a single read canister call.
+    max_entries_per_call: u16,
+    /// Timestamp of the last install / upgrade of this canister.
+    last_upgrade_timestamp: Timestamp,
+}
+
+impl Storable for ConfigState {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        match &self {
+            ConfigState::Uninitialized => Cow::Borrowed(&[]),
+            ConfigState::Initialized(config) => {
+                let buf = candid::encode_one(config).expect("failed to encode archive config");
+                Cow::Owned(buf)
+            }
+        }
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        if bytes.len() == 0 {
+            return ConfigState::Uninitialized;
+        }
+        ConfigState::Initialized(
+            candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode archive config"),
+        )
+    }
+}
+
+/// Index key for the anchor index.
+/// Changing the (serialized) size of this value requires a stable memory migration.
```

Comment:
What would happen without migration? Would this be caught by the type system or by tests?

Reply:
I added unit tests that catch unexpected changes.